tutorials/graph-theory/binary-search-tree.md

```diff
@@ -1,12 +1,304 @@
+3. If our target node value matches with the value of current node then return that node.
+4. If value not matched with current node then we either procced to right or left child according to the condition.
+5. If target node value is greater than the current node then we procced towards the right child or right subtree.
+6. If target node value is lower than the current node then we proceed towards the left child or left subtree.
+
+`target value: value of node you want to search in tree`
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
+// For BST search
+Node *bstSearch(int target,Node* root){
+    if(root == NULL)
+        return root;
+    if(root->data == target)
+        return root;
+    if(root->data<target){
+      // search target node in right subtree
+        return bstSearch(target,root->right);
+    } else{
+       // search target node in left subtree
+        return bstSearch(target,root->left);
+    }
+}
+```
+
+</TabItem>
+</Tabs>
+
+### 2. Insertion
+
+**Algorithm**
+
+1. Insertion begin with searh operation.
+2. Search left or right subtree according to the relation between node value and the value of our target node as we did in search.
+3. Repeat step 1 until we reach to correct place to insert our new node.
+4. Add a new node as a right or left child according to the node value and our target node value.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
+// For BST insertion
+Node *bstInsertion(int val,Node *node){
+    if(node==NULL){
+      // Insert new node
+        node = new Node(val);
+    } else if(node->data>val){
+        // Search correct place to insert in left subtree
+        node->left = bstInsertion(val,node->left);
+    } else if(node->data<val){
+      // Search correct place to insert in right subtree
+        node->right = bstInsertion(val,node->right);
+    }
+    return node;
+}
+```
+
+</TabItem>
+</Tabs>
+
+### 3. Deletion
+
+**Algorithm**
+
+In order to delete any node in BST there are some scenarios we need to cover
+
+1. First of all we need to search that node in BST. Same step as we followed above for search.
+2. Now there are three cases for node we got after search that needs to be deleted.
+
+**Case 1**
+
+If the node to be deleted is a leaf node. In such case simply delete that node from tree.
+
+**Case 2**
+
+The node to be deleted has one child node. Then in that case:
+
+- Replace that node with it's child node.
+- Delete original child node from the tree.
+
+**Case 3**
+
+The node to be deleted has two child node. Then in that case:
+
+- Find the inorder successor of that node. (Inorder successor:- the leftmost node in the right subtree).
+- Replace that node with inorder successor.
+- Delete inorder successor from it's original position.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@utkarsh"/>
+
+```cpp
+Node *inorderSuccessor(Node* root){
```

Comment:
Format your code with K&R style.

See https://gist.github.com/jesseschalken/0f47a2b5a738ced9c845#why-kr for example